### PR TITLE
make binary: do not ignore unresolved symbols

### PR DESCRIPTION
**Pull request by kolyshkin**
TL;DR: stop building static binary that may fail

Linker flag --unresolved-symbols=ignore-in-shared-libs was added
in commit 06d0843 two years ago for the static build case, presumably
to avoid dealing with problem of missing libraries.

For the record, this is what ld(1) man page says:

```
--unresolved-symbols=method
   Determine how to handle unresolved symbols.  There are four
   possible values for method:
        .........
    ignore-in-shared-libs
       Report unresolved symbols that come from regular object files,
       but ignore them if they come from shared libraries.  This can
       be useful when creating a dynamic binary and it is known that
       all the shared libraries that it should be referencing are
       included on the linker's command line.
```

Here, the flag is not used for its purpose ("creating a dynamic binary")
and does more harm than good. Instead of complaining about missing symbols
as it should do if some libraries are missing from LIBS/LDFLAGS, it lets
ld create a binary with unresolved symbols, ike this:

```
 $ readelf -s bundles/1.7.1/binary/docker-1.7.1 | grep -w UND
 ........
 21029: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND dlopen
 .........
```

Such binary is working just fine -- until code calls one of those
functions, then it crashes (for apparently no reason, i.e. it is
impossible to tell why from the diagnistics printed).

In other words, adding this flag allows to build a static binary
with missing libraries, hiding the problem from both a developer
(who forgot to add a library to #cgo: LDFLAGS -- I was one such
developer a few days ago when I was working on ploop graphdriver)
and from a user (who expects the binary to work without crashing,
and it does that until the code calls a function in one of those
libraries).

Removing the flag immediately unveils the problem (as it should):

```
    /usr/lib/gcc/x86_64-linux-gnu/4.8/../../../x86_64-linux-gnu/libsqlite3.a(sqlite3.o):
    In function `unixDlError':
    (.text+0x20971): undefined reference to `dlerror'
    /usr/lib/gcc/x86_64-linux-gnu/4.8/../../../x86_64-linux-gnu/libsqlite3.a(sqlite3.o):
    In function `unixDlClose':
    (.text+0x8814): undefined reference to `dlclose'
```

The problem is, gosqlite package says:

```
    #cgo LDFLAGS: -lsqlite3
```

which is enough for dynamic linking, as indirect dependencies (i.e.
libraries required by libsqlite3.so) are listed in .so file and will be
resolved dynamically by ldd upon executing the binary.

For static linking though, one has to list all the required libraries,
both direct and indirect. For libraries with pkgconfig support the
list of required libraries can be obtained with pkg-config:

```
    $ pkg-config --libs sqlite3 # dynamic linking case
    -lsqlite3
    $ pkg-config --libs --static sqlite3 # static case
    -lsqlite3 -ldl -lpthread
```

It seems that all one has to do is to fix gosqlite this way:

``` udiff
    -#cgo LDFLAGS: -lsqlite3
    +#cgo pkg-config: sqlite3
```

Unfortunately, cmd/go doesn't know that it needs to pass --static flag to pkg-config in case of static linking (see https://github.com/golang/go/issues/12058).

So, for one, one has to do one of these things:

1 Patch sqlite.go like this:

``` udiff
    -#cgo LDFLAGS: -lsqlite3
    +#cgo pkg-config: --static sqlite3
```

(this is exactly what I do in goploop, see https://github.com/kolyshkin/goploop/commit/e9aa072f51)

2 Patch sqlite.go like this:

``` udiff
    -#cgo LDFLAGS: -lsqlite3
    +#cgo LDFLAGS: -lsqlite3 -ldl -lpthread
```

(I would submit this patch to gosqlite but it seems that
https://code.google.com/p/gosqlite/ is deserted and not maintained,
and patching it here is not right as it is "vendored")

3 Explicitly add `-ldl` for the static link case.
This is what this patch does.

4 Fork sqlite to github and maintain it there. Personally I am not
ready for that, as I'm neither a Go expert nor gosqlite user.

Now, solution this commit uses (3) doesn't look like a clear one, but nevertheless it makes
the build much better than it was before.

Signed-off-by: Kir Kolyshkin kir@openvz.org

### CODE REVIEW

**Comment by LK4D4** [comment]
@tianon aha, that's why we saw this for new sqlite lib and not for old, because we're suppressed warning. And I think we merged crappy patch with two sqlites only because of this.
@kolyshkin Thanks for research! I think it'll be possible to replace old gosqlite with github.com/mattn/go-sqlite3 which was used for windows anyway.
LGTM for now

**Comment by LK4D4** [comment]
Oh no, I'm sorry, we saw another warning, but I'm not sure if it can be helped:
```
/usr/lib/gcc/x86_64-linux-gnu/4.8/../../../x86_64-linux-gnu/libsqlite3.a(sqlite3.o): In function `unixDlOpen':
(.text+0x8879): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
```

**Comment by kolyshkin** [comment]
This warning is expected; the flag this patch removes was just hiding it, not fixing anything. The only way to fix it for real is to recompile libsqlite without dlopen support (I don't remember what flag is it).

**Comment by LK4D4** [comment]
@tianon @tiborvass @jhowardmsft if we anyway will have this warning, maybe makes sense to move to mattn/go-sqlite3. I can try to compile sqlite from sources without `dlopen` in Dockerfile if it's needed.

**Comment by jhowardmsft** [comment]
@LK4D4 I think a lot of people would be happier if you can get mattn/go-sqlite3 working on Linux. As you know, quite a few people have failed now...

**Comment by LK4D4** [comment]
@jhowardmsft I tried to compile it and it worked for me. What was actually the problem?

**Comment by jhowardmsft** [comment]
https://github.com/docker/docker/pull/13617 comments. It might be the same as this.

**Comment by LK4D4** [comment]
@jhowardmsft Yes, all worked okay just for old sqlite we suppressed warning.

**Comment by kolyshkin** [comment]
for github.com/mattn/go-sqlite3, removing ```dlopen()``` and friends from the binary is as easy is
```udiff
diff --git a/sqlite3.go b/sqlite3.go
index d57d9fb..11d4a80 100644
--- a/sqlite3.go
+++ b/sqlite3.go
@@ -9,6 +9,7 @@ package sqlite3
 #cgo CFLAGS: -std=gnu99
 #cgo CFLAGS: -DSQLITE_ENABLE_RTREE -DSQLITE_THREADSAFE
 #cgo CFLAGS: -DSQLITE_ENABLE_FTS3 -DSQLITE_ENABLE_FTS3_PARENTHESIS
+#cgo CFLAGS: -DSQLITE_OMIT_LOAD_EXTENSION
 #include <sqlite3-binding.h>
 #include <stdlib.h>
 #include <string.h>
```
The two questions are:
Q1: Is anyone is using this functionality?
A1: Apparently not in Docker, otherwise it would crash as I described
Q2: Is it possible to do it conditionally, like for static linking only?
A2: I don't know a way to do that.

**Comment by kolyshkin** [comment]
@jhowardmsft if you mean this
> libsqlite3.a(sqlite3.o): In function `unixDlOpen':
> (.text+0x8879): warning: Using 'dlopen' in statically linked applications requires
> at runtime the shared libraries from the glibc version used for linking
then
1 it's just a warning not an error
2 it's not preventing the library from working, except maybe for the "extension loading" functionality which is apparently not in use by Docker
3 it's not a regression, actually quite the opposite (it was silent fatal runtime error, now it's a compile time warning)

**Comment by kolyshkin** [comment]
In other words, before this patch, if anyone would use the SQLite ```load_extention()``` feature (https://www.sqlite.org/loadext.html) from a statically linked Docker binary, the binary would surely crash with SIGSEGV due to unresolved dlopen() symbol. With this patch, the SQLite ```load_extension()``` _might_ not work, plus we get a linker warning telling this.

**Comment by LK4D4** [comment]
btw, from sqlite site:
```
The use of the amalgamation is recommended for all applications.
```

**Comment by kolyshkin** [comment]
Apparently we're good, as one doesn't need "re-amalgamation" for SQLITE_OMIT_LOAD_EXTENSION

**Comment by kolyshkin** [comment]
@LK4D4 So, what are you going to do about it? It makes sense to switch to @mattn go-sqlite3 and either ignore the warning for the time being, or maybe ask @mattn to add a patch like I propose earlier (frankly I don't know how many users does this LOAD_EXTENSION feature has, but I suspect not too many).

Also, I guess, adding a flag to CFLAGS can be implemented via tags, albeit it will probably look ugly. I kinda miss #ifdef :)

**Comment by calavera** [comment]
Is there any resolution here? @kolyshkin can we close this PR?

**Comment by kolyshkin** [comment]
@calavera I don't have much to add to what I have already explained in great details above, so let me try to express it differently.

I strongly believe you should merge this patch, because otherwise you have an option to linker that silently ignores unresolved symbols. The consequences are:
* hiding important and not-so-important warnings
* possible creation of binaries that might crash
* developer frustration (took me more than a day to figure out I forgot to add a library)

The only consequence of merging this patch I know is a (legitimate) linker warning.

**Comment by tianon** [comment]
@lk4d4 the reason we weren't able to use go-sqlite3 universally was the
embedded sqlite, which he did add a build flag for, but I haven't had a
chance to test to make sure that'll be enough for packaging yet

**Comment by tianon** [comment]
Do you think `#cgo CFLAGS: -DSQLITE_OMIT_LOAD_EXTENSION` could be added via build tags on go-sqlite3?

**Comment by LK4D4** [comment]
@tianon I don't believe it helps with using system sqlite3

**Comment by tianon** [comment]
Naw, that one's for static compilation (which is something we need, but not many other consumers of go-sqlite3 do).

I think it's worth noting that someone else packaged go-sqlite3 for Debian, and they did so by patching out the build flag entirely and using that file all the time, which is a decent solution IMO, so I'm satisfied with switching if we make sure it's on-par (although @cpuguy83 is gunning to just remove SQLite entirely :smiling_imp:).

**Comment by jfrazelle** [comment]
I have opened this: https://github.com/mattn/go-sqlite3/pull/235

**Comment by calavera** [comment]
I'm closing this PR because the changes here will be included with #16094. Thank you!
